cbbpoll/views.py

```diff
@@ -282,8 +283,20 @@ def submitballot():
         else:
             ballot = Ballot(updated = datetime.utcnow(), poll_id = poll.id, user_id = g.user.id)
         db.session.add(ballot)
-        # must commit to get ballot id
-        db.session.commit()
+
+        # must commit here to get ballot id for the Vote objects.
+        #
+        # there is a race condition here with the check to see if a ballot exists earlier in this function.
+        # the database maintains the invariant of one ballot per (user, poll) pair, so this commit may fail.
+        # if it does, it's probably alright, as it means we came here with no ballot existing, but now one exists.
+        # so send them to their most recent ballot.
+
+        try:
+            db.session.commit()
+        except IntegrityError:
+            db.session.rollback()
+            return redirect(url_for('ballot'))
```

Comment:
Decided to grab the ballot.  Thought it would be more work than it was--realized we already do it above (it's the entrance condition for the race, after all).  Might as well save the redirect.

Left /ballot in as a convenience route--it redirects so it doesn't break the fact that right now essentially all of our URLs (in the address bar after page load) are permalinks.